File: analysis/energyday.py
from datetime import datetime

from sunsynk.resource import Resource
from analysis.energysummary import QueryType
from analysis.virtualbattery import VirtualBattery
from analysis.energymonth import EnergyMonth


class IntervalSummary(Resource):
    """Accumulates peak/offpeak Wh totals from 5-minute interval records for one label."""

    def __init__(self, data, battery: VirtualBattery,
                 offpeakstart="00:00", offpeakstop="00:07", is_load: bool = False):
        self.label = data['label']
        self.records = data['records']
        self.peak = 0
        self.peakexport = 0
        self.offpeak = 0
        self.offpeakexport = 0
        self.offpeakpercentage = 0
        is_off_peak = False
        recharged = False
        battery_ran_out = False

        start = datetime.strptime(offpeakstart, "%H:%M")
        stop = datetime.strptime(offpeakstop, "%H:%M")

        for record in self.records:
            time = datetime.strptime(record['time'], "%H:%M")
            value = float(record['value']) / 12

            if time >= start:
                is_off_peak = True
            if is_off_peak:
                if time >= stop:
                    is_off_peak = False

            if is_off_peak:
                if not recharged:
                    recharged = True
                    if is_load:
                        battery.recharge()

                if value > 0:
                    self.offpeak = self.offpeak + value
                else:
                    extra_load = min(value * -1, 2.3)
                    export = (value * -1) - extra_load
                    self.offpeakexport = self.offpeakexport + export
                    self.offpeak = self.offpeak - extra_load
            else:
                if value > 0:
                    self.peak = self.peak + value
                    if is_load:
                        temp = battery.utilise(value, time)
                        if temp > 0:
                            battery_ran_out = True
                else:
                    extra_load = min(value * -1, 2.3)
                    export = (value * -1) - extra_load
                    self.peakexport = self.peakexport + export
                    self.peak = self.peak - extra_load
                    if is_load:
                        battery.pv_charge(export, time)

        if battery_ran_out:
            battery.set_ran_out()

        if self.offpeak + self.peak > 0:
            self.offpeakpercentage = self.offpeak / (self.offpeak + self.peak)


class EnergyDay(Resource):
    def __init__(self, data, date: str, month: EnergyMonth, battery: VirtualBattery,
                 offpeakstart: str, offpeakstop: str):
        self.data = data
        energy = self.data['infos']
        for item in energy:
            if item['label'] == "PV":
                self.pv = IntervalSummary(item, battery, offpeakstart, offpeakstop)
            elif item['label'] == "Grid":
                self.grid = IntervalSummary(item, battery, offpeakstart, offpeakstop, is_load=True)
            elif item['label'] == "Load":
                self.load = IntervalSummary(item, battery, offpeakstart, offpeakstop)

        self.supplied_load = next((float(r['value']) for r in month.get_load()['records'] if r['time'] == date), 0.0)
        self.supplied_import = next((float(r['value']) for r in month.get_import()['records'] if r['time'] == date), 0.0)
        self.supplied_pv = next((float(r['value']) for r in month.get_pv()['records'] if r['time'] == date), 0.0)
        self.supplied_export = next((float(r['value']) for r in month.get_export()['records'] if r['time'] == date), 0.0)

    def get_calc_export(self, qtype: QueryType = QueryType.BOTH):
        if qtype == QueryType.BOTH:
            return self.grid.peakexport + self.grid.offpeakexport
        if qtype == QueryType.PEAK:
            return self.grid.peakexport
        if qtype == QueryType.OFFPEAK:
            return self.grid.offpeakexport
        return 0

    def get_calc_import(self, qtype: QueryType = QueryType.BOTH):
        if qtype == QueryType.BOTH:
            return self.grid.peak + self.grid.offpeak
        if qtype == QueryType.PEAK:
            return self.grid.peak
        if qtype == QueryType.OFFPEAK:
            return self.grid.offpeak
        return 0

    def get_calc_export_peak(self):
        return self.grid.peakexport

    def get_calc_import_peak(self):
        return self.grid.peak

    def get_calc_export_off_peak(self):
        return self.grid.offpeakexport

    def get_calc_import_off_peak(self):
        return self.grid.offpeak

    def get_calc_pv(self):
        return self.pv.peak + self.pv.offpeak

    def get_calc_pv_peak(self):
        return self.pv.peak

    def get_calc_pv_off_peak(self):
        return self.pv.offpeak

    def get_calc_load(self):
        return self.load.peak + self.load.offpeak

    def get_calc_load_peak(self):
        return self.load.peak

    def get_calc_load_off_peak(self):
        return self.load.offpeak

    def get_supplied_load(self):
        return self.supplied_load

    def get_supplied_export(self):
        return self.supplied_export

    def get_supplied_import(self):
        return self.supplied_import

    def get_supplied_pv(self):
        return self.supplied_pv

    def print(self):
        if self.get_calc_export(QueryType.BOTH) > 0:
            print(f"Export: {round(self.get_calc_export() / 1000, 2)}kWh (vs {round(self.get_supplied_export(), 1)}kWh), "
                  f"Diff = {round(self.get_supplied_export() - self.get_calc_export() / 1000, 2)}kWh  "
                  f"%age {round(self.get_supplied_export() / (self.get_calc_export() / 1000), 2)}")
        print(f"Import {round(self.get_calc_import() / 1000, 2)}kWh (vs {round(self.get_supplied_import(), 1)}kWh), "
              f"Diff = {round(self.get_supplied_import() - self.get_calc_import() / 1000, 2)}kWh")
        print(f"PV {round(self.get_calc_pv() / 1000, 2)}kWh (vs {round(self.get_supplied_pv(), 1)}kWh), "
              f"Diff = {round(self.get_supplied_pv() - self.get_calc_pv() / 1000, 2)}kWh   "
              f"%age {round(self.get_supplied_pv() / (self.get_calc_pv() / 1000), 2)}")

File: android/app/src/main/java/com/richw/sunsynk/analysis/EnergyDay.kt
package com.richw.sunsynk.analysis

import com.google.gson.JsonObject
import java.time.LocalTime
import java.time.format.DateTimeFormatter

class IntervalSummary(
    data: JsonObject,
    battery: VirtualBattery?,
    offPeakStart: String,
    offPeakStop: String,
    private val isLoad: Boolean = false,
) {
    val label: String = data.get("label").asString
    var peak: Double = 0.0
    var peakExport: Double = 0.0
    var offPeak: Double = 0.0
    var offPeakExport: Double = 0.0
    var offPeakPercentage: Double = 0.0

    val parsed: MutableList<Pair<LocalTime, Double>> = if (isLoad) mutableListOf() else mutableListOf()

    init {
        val fmt = DateTimeFormatter.ofPattern("H:mm")
        val start = LocalTime.parse(offPeakStart, fmt)
        val stop = LocalTime.parse(offPeakStop, fmt)
        var recharged = false
        var batteryRanOut = false

        val records = data.getAsJsonArray("records")
        for (r in records) {
            val rec = r.asJsonObject
            val time = LocalTime.parse(rec.get("time").asString, fmt)
            val value = rec.get("value").asString.toDouble() / 12.0
            if (isLoad) parsed.add(Pair(time, value))

            if (!time.isBefore(start) && time.isBefore(stop)) {
                recharged = processOffPeak(value, battery, isLoad, recharged)
            } else {
                if (processPeak(value, time, battery, isLoad)) batteryRanOut = true
            }
        }

        if (battery != null && batteryRanOut) battery.setRanOut()

        val total = offPeak + peak
        if (total > 0) offPeakPercentage = offPeak / total
    }

    private fun processOffPeak(value: Double, battery: VirtualBattery?, isLoad: Boolean, recharged: Boolean): Boolean {
        if (!recharged) {
            if (battery != null && isLoad) battery.recharge()
        }
        if (value > 0) {
            offPeak += value
        } else {
            val extraLoad = minOf(-value, 2.3)
            offPeakExport += -value - extraLoad
            offPeak -= extraLoad
        }
        return true
    }

    private fun processPeak(value: Double, time: LocalTime, battery: VirtualBattery?, isLoad: Boolean): Boolean {
        return if (value > 0) {
            peak += value
            if (battery != null && isLoad) {
                battery.utilise(value, time) > 0
            } else false
        } else {
            val extraLoad = minOf(-value, 2.3)
            val export = -value - extraLoad
            peakExport += export
            peak -= extraLoad
            if (battery != null && isLoad) {
                battery.pvCharge(export, time)
            }
            false
        }
    }
}

class EnergyDay(
    data: JsonObject,
    date: String,
    month: EnergyMonth,
    battery: VirtualBattery?,
    offPeakStart: String,
    offPeakStop: String,
) {
    val pv: IntervalSummary
    val grid: IntervalSummary
    val load: IntervalSummary
    val suppliedLoad: Double
    val suppliedImport: Double
    val suppliedPv: Double
    val suppliedExport: Double

    private val startDt: LocalTime
    private val stopDt: LocalTime

    init {
        val fmt = DateTimeFormatter.ofPattern("H:mm")
        startDt = LocalTime.parse(offPeakStart, fmt)
        stopDt = LocalTime.parse(offPeakStop, fmt)

        val infos = data.getAsJsonArray("infos")
        var pvTmp: IntervalSummary? = null
        var gridTmp: IntervalSummary? = null
        var loadTmp: IntervalSummary? = null
        for (item in infos) {
            val obj = item.asJsonObject
            when (obj.get("label").asString) {
                "PV" -> pvTmp = IntervalSummary(obj, battery, offPeakStart, offPeakStop)
                "Grid" -> gridTmp = IntervalSummary(obj, battery, offPeakStart, offPeakStop, isLoad = true)
                "Load" -> loadTmp = IntervalSummary(obj, battery, offPeakStart, offPeakStop)
            }
        }
        pv = pvTmp ?: IntervalSummary(emptyInfoObj("PV"), null, offPeakStart, offPeakStop)
        grid = gridTmp ?: IntervalSummary(emptyInfoObj("Grid"), null, offPeakStart, offPeakStop, isLoad = true)
        load = loadTmp ?: IntervalSummary(emptyInfoObj("Load"), null, offPeakStart, offPeakStop)

        suppliedLoad = month.getLoad()?.records?.firstOrNull { it.time == date }?.value?.toDoubleOrNull() ?: 0.0
        suppliedImport = month.getImport()?.records?.firstOrNull { it.time == date }?.value?.toDoubleOrNull() ?: 0.0
        suppliedPv = month.getPv()?.records?.firstOrNull { it.time == date }?.value?.toDoubleOrNull() ?: 0.0
        suppliedExport = month.getExport()?.records?.firstOrNull { it.time == date }?.value?.toDoubleOrNull() ?: 0.0
    }

    fun runBattery(battery: VirtualBattery) {
        var recharged = false
        var batteryRanOut = false
        for ((timeDt, wh) in grid.parsed) {
            if (!timeDt.isBefore(startDt) && timeDt.isBefore(stopDt)) {
                if (!recharged) {
                    battery.recharge()
                    recharged = true
                }
            } else {
                if (wh > 0) {
                    if (battery.utilise(wh, timeDt) > 0) batteryRanOut = true
                } else {
                    val extraLoad = minOf(-wh, 2.3)
                    battery.pvCharge(-wh - extraLoad, timeDt)
                }
            }
        }
        if (batteryRanOut) battery.setRanOut()
    }

    fun getCalcExport() = grid.peakExport + grid.offPeakExport
    fun getCalcImport() = grid.peak + grid.offPeak
    fun getCalcExportPeak() = grid.peakExport
    fun getCalcImportPeak() = grid.peak
    fun getCalcExportOffPeak() = grid.offPeakExport
    fun getCalcImportOffPeak() = grid.offPeak
    fun getCalcPv() = pv.peak + pv.offPeak
    fun getCalcLoad() = load.peak + load.offPeak
    fun getCalcLoadPeak() = load.peak
    fun getCalcLoadOffPeak() = load.offPeak
}

private fun emptyInfoObj(label: String): JsonObject {
    val obj = JsonObject()
    obj.addProperty("label", label)
    obj.add("records", com.google.gson.JsonArray())
    return obj
}
